File: day4/solution.py
# problem statement at https://adventofcode.com/2019/day/4

def check(password: str) -> bool:
    increasing = True
    has_double = False

    last_char = password[0]
    min_num = int(password[0])

    for char in password[1:]:
        if char == last_char:
            has_double = True
        
        if int(char) < min_num:
            increasing = False
        
        last_char = char
        min_num = int(char)
    
    return has_double and increasing

def check_2(password: str) -> bool:
    increasing = True
    has_double = False

    last_char = password[0]
    min_num = int(password[0])

    consecutive_number = 1

    for char in password[1:]:
        if char == last_char:
            consecutive_number += 1
        else:
            if consecutive_number == 2:
                has_double = True
            else:
                consecutive_number = 1

        if int(char) < min_num:
            increasing = False
        
        last_char = char
        min_num = int(char)
    
    if consecutive_number == 2:
        has_double = True

    return has_double and increasing

if __name__ == '__main__':
    puzzle_input = '125730-579381'

    start, end = puzzle_input.split('-')

    valid_passwords = 0
    valid_passwords_2 = 0

    for num in range(int(start), int(end)+1):
        if check(str(num)):
            valid_passwords += 1
        if check_2(str(num)):
            valid_passwords_2 += 1
    
    print(valid_passwords, valid_passwords_2)

File: day4/solution.ts
import { logger } from "../utils/logger";

function validPassword(password: string): boolean {
  let increasing = true;
  let hasDouble = false;

  let lastChar = password.charAt(0);
  let maxNum = parseInt(password.charAt(0));
  for (let char of password.slice(1)) {
    const numericVal = parseInt(char);

    if (numericVal < maxNum) {
      increasing = false;
    } else {
      maxNum = numericVal;
    }

    if (char == lastChar) {
      hasDouble = true;
    }
    lastChar = char;
  }

  return increasing && hasDouble;
}

function validPasswordForPart2(password: string): boolean {
  let increasing = true;
  let hasDouble = false;

  let lastChar = password.charAt(0);
  let maxNum = parseInt(password.charAt(0));
  let consecutiveChars = 1;
  for (let char of password.slice(1)) {
    const numericVal = parseInt(char);

    if (numericVal < maxNum) {
      increasing = false;
    } else {
      maxNum = numericVal;
    }

    if (char == lastChar) {
      consecutiveChars++;
    } else {
      if (consecutiveChars == 2) {
        hasDouble = true;
      }
      consecutiveChars = 1;
    }
    lastChar = char;
  }

  if (consecutiveChars == 2) {
    hasDouble = true;
  }

  return increasing && hasDouble;
}

const puzzleInput = '125730-579381';
const input = puzzleInput.split('-').map(str => parseInt(str.trim()));
logger.debug(input);

const [start, end] = input;
logger.debug(start, end);

let answer = 0;
let answerPart2 = 0;
for (let i = start; i <= end; i++) {
  logger.verbose('checking', i);
  if (validPassword(i.toString())) {
    logger.progress(i, 'is valid for part 1');
    answer++;
  }

  if (validPasswordForPart2(i.toString())) {
    logger.progress(i, 'is valid for part 2');
    answerPart2++;
  }
}
logger.solution('part 1', answer);
logger.solution('part 2', answerPart2);
